okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/OkHttpClientsTest.java

```diff
@@ -125,6 +133,29 @@ public void doesNotShareBackoffStateBetweenCalls() throws Exception {
         assertThat(server.getRequestCount()).isEqualTo(4 /* two from each call */);
     }
 
+    @Test
+    public void interceptsAndHandlesQosIoExceptions_endToEnd_asyncCall() throws Exception {
+        server.enqueue(new MockResponse().setResponseCode(503));
+        server.enqueue(new MockResponse().setResponseCode(503));
+        server.enqueue(new MockResponse().setBody("pong"));
+
+        Call call = createRetryingClient(2).newCall(new Request.Builder().url(url).build());
+        CompletableFuture<String> future = new CompletableFuture<>();
+        call.enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException ioException) {
+                future.completeExceptionally(ioException);
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                future.complete(response.body().string());
+            }
+        });
+        assertThat(future.get()).isEqualTo("pong");
```

Comment:
Generally best to use the calls that time out, as otherwise a code failure becomes a hard-to-diagnose hung test

Reply:
good idea.